### ISSUE TITLE
allow dynamic port binding for native apps in end session

### ISSUE THREAD

**Issue by ahmednfwela** [zitadel/zitadel#6615]
## User story

Allow dynamic port binding for native apps that use loop back addresses. The behavior is outlined [RFC8252, section 7.3](https://datatracker.ietf.org/doc/html/rfc8252#section-7.3)  but normally only applies to Authorization requests within the Oauth2 universe.

End-session is an OIDC-only endpoint specified in [OpenID Connect RP-Initiated Logout 1.0](https://openid.net/specs/openid-connect-rpinitiated-1_0.html) and therefore we did not take RFC8252 into account for the `post_logout_redirect_uri` parameter.

### Acceptance criteria

- [ ] Implement [RFC8252, section 7.3](https://datatracker.ietf.org/doc/html/rfc8252#section-7.3) for the `post_logout_redirect_uri` parameter in the End-Session endpoint
- [ ] Submit a suggestion to the OpenID foundation to clarify the RP Initiated Logout specifications wrt to Native apps.

----

### Preflight Checklist

- [X] I could not find a solution in the documentation, the existing issues or discussions
- [X] I have joined the [ZITADEL chat](https://zitadel.com/chat)

### Environment

ZITADEL Cloud

### Describe the problem caused by this bug

we get post_logout_redirect_uri invalid when calling end_session endpoint for native apps.

### To reproduce

when adding a native (desktop) app to zitadel, it asks for the redirect urls and the post logout redirect url.
we set them both to the same url: `http://127.0.0.1`.
this works fine for authentication, but when we logout using the `end_session_endpoint`, we get 
```json
{
  "error": "invalid_request",
  "error_description": "post_logout_redirect_uri invalid"
}
```
when we pass the uri as: `http://127.0.0.1:12345` where `12345` is a random available port.

**Comment by hifabienne** [comment]
I think that the url you register in ZITADEL needs to match exactly what you are sending.
@livio-a @muhlemmer is that right?

**Comment by ahmednfwela** [comment]
@hifabienne but that isn't the case for the auth request, the bug only exists for the logout request

**Comment by muhlemmer** [comment]
The complete URL must exactly match the settings in Redirect URI and Logout Redirect URI. Random ports won't work. Can you send and example payload that is send to the End Session endpoint? It can also be that `localhost` and `127.0.0.1` are mixed up. Although they resolve to the same server (most of the time) the host names are not considered equal from a OIDC perspective.

**Comment by ahmednfwela** [comment]
@muhlemmer per [the spec](https://datatracker.ietf.org/doc/html/rfc8252#section-7.3), random ports SHOULD work

**Comment by muhlemmer** [comment]
> @muhlemmer per [the spec](https://datatracker.ietf.org/doc/html/rfc8252#section-7.3), random ports SHOULD work

Interesting find. [RFC8252, section 7.3](https://datatracker.ietf.org/doc/html/rfc8252#section-7.3) directly contradicts the OpenID connect standard that we implement. We maked no claim RFC8252 is implemented in zitadel or our OIDC library. The Oauth and OpenID specs we currently implement are [listed here](https://github.com/zitadel/oidc?tab=readme-ov-file#features).

----

OpenID on `redirect_uri`  in the Auth Request: [`redirect_uri`](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest) 

> redirect_uri
> REQUIRED. Redirection URI to which the response will be sent. This URI MUST exactly match one of the Redirection URI values for the Client pre-registered at the OpenID Provider, with the matching performed as described in Section 6.2.1 of [RFC3986](https://openid.net/specs/openid-connect-core-1_0.html#RFC3986) (Simple String Comparison). When using this flow, the Redirection URI SHOULD use the https scheme; however, it MAY use the http scheme, provided that the Client Type is confidential, as defined in Section 2.1 of OAuth 2.0, and provided the OP allows the use of http Redirection URIs in this case. Also, if the Client is a native application, it MAY use the http scheme with localhost or the IP loopback literals 127.0.0.1 or [::1] as the hostname. The Redirection URI MAY use an alternate scheme, such as one that is intended to identify a callback into a native application.

OpenID RP-initiated logout on [redirection](https://openid.net/specs/openid-connect-rpinitiated-1_0.html#RedirectionAfterLogout):

> The OP also MUST NOT perform post-logout redirection if the post_logout_redirect_uri value supplied does not exactly match one of the previously registered post_logout_redirec749t_uris values

And:

> post_logout_redirect_uris
> OPTIONAL. Array of URLs supplied by the RP to which it MAY request that the End-User's User Agent be redirected using the post_logout_redirect_uri parameter after a logout has been performed. These URLs SHOULD use the https scheme and MAY contain port, path, and query parameter components; however, they MAY use the http scheme, provided that the Client Type is confidential, as defined in Section 2.1 of [OAuth 2.0](https://openid.net/specs/openid-connect-rpinitiated-1_0.html#RFC6749) [RFC6749], and provided the OP allows the use of http RP URIs.

RFC6749 makes no further mentions on dynamic ports that may be used for native apps. [Section 3.1.2.3](https://www.rfc-editor.org/rfc/rfc6749.html#section-3.1.2.3) is consistent with the OpenID spec above.

----

To summarize: @ahmednfwela expects RFC8252 which is currently not fully implemented. RFC8252 updates RFC6749 (Oauth2), but the update is not incorporated in OpenID connect (yet).

To be investigated:
- Is there a issue open at the OpenID connect foundation? If not, open one and ask for clarification.
- Should we implement RFC8252, what are pros, cons and costs. How many people need this?

IMO this is a feature request, not bug.

CC @livio-a, @hifabienne

**Comment by ahmednfwela** [comment]
@muhlemmer Since zitadel gives the option to select a native app 
![image](https://github.com/zitadel/zitadel/assets/63286031/4b88dff7-94c2-4839-a904-632b20df137b)

the first thing that came to mind was RFC8252, I didn't expect zitadel to support desktop without allowing random ports

**Comment by muhlemmer** [comment]
I understand and your expectations are reasonable. For us it is not a simple fix to switch from simple string compare to parse-and-exclude port. Plus it is a behaviour change. Hence, I prefer to plan it as a feature.

**Comment by muhlemmer** [comment]
I've modified the original issue with acceptance criteria for a feature. This will now be moved to the backlog and based on demand we will prioritize it. Meanwhile, PRs are welcome.

**Comment by ahmednfwela** [comment]
> * `post_logout_redirect_uri` is part of the [OpenID Connect RP-Initiated Logout 1.0](https://openid.net/specs/openid-connect-rpinitiated-1_0.html) spec, which only references RFC6749, basic Oauth2.
> * The End Session endpoint is not an Oauth2 endpoint but a OIDC-only endpoint for RP-initiated logout.

I would disagree with the notion that RFC6749 and RFC8252 are 2 different things, since RFC8252  is explicitly stated as an update for RFC6749 
![Image](https://github.com/user-attachments/assets/a2fed91d-a8e7-4e66-8fe9-f696fec855ec)

also now that #8287 is supposed to be finished, will it support logging out of native apps ?

**Comment by ahmednfwela** [comment]
@muhlemmer I have opened https://github.com/zitadel/oidc/pull/875 to fix this issue

Hope this helps push this fix faster, thanks!
